src/java.base/share/classes/java/time/Clock.java

```diff
@@ -473,32 +451,120 @@ public  int hashCode() {
         return super.hashCode();
     }
 
+    //-----------------------------------------------------------------------
+    // initial offset
+    private static final long OFFSET_SEED = System.currentTimeMillis() / 1000 - 1024;
+    // We don't actually need a volatile here.
+    // We don't care if offset is set or read concurrently by multiple
+    // threads - we just need a value which is 'recent enough' - in other
+    // words something that has been updated at least once in the last
+    // 2^32 secs (~136 years). And even if we by chance see an invalid
+    // offset, the worst that can happen is that we will get a -1 value
+    // from getNanoTimeAdjustment, forcing us to update the offset
+    // once again.
+    private static long offset = OFFSET_SEED;
+
+    static Instant currentInstant() {
+        // Take a local copy of offset. offset can be updated concurrently
+        // by other threads (even if we haven't made it volatile) so we will
+        // work with a local copy.
+        long localOffset = offset;
+        long adjustment = VM.getNanoTimeAdjustment(localOffset);
+
+        if (adjustment == -1) {
+            // -1 is a sentinel value returned by VM.getNanoTimeAdjustment
+            // when the offset it is given is too far off the current UTC
+            // time. In principle, this should not happen unless the
+            // JVM has run for more than ~136 years (not likely) or
+            // someone is fiddling with the system time, or the offset is
+            // by chance at 1ns in the future (very unlikely).
+            // We can easily recover from all these conditions by bringing
+            // back the offset in range and retry.
+
+            // bring back the offset in range. We use -1024 to make
+            // it more unlikely to hit the 1ns in the future condition.
+            localOffset = System.currentTimeMillis()/1000 - 1024;
+
```

Comment:
This isn't my logic - it is existing code that has been moved. I'm not a fan of infinite retry loops as the can hang the system. But I'm happy to change it to work that way if there is a consensus to do so.

Reply:
I see the localOffset passed to VM.getNanoTimeAdjustment() has to be in the range of current time +/- 2^32 seconds. This means the thread would have to sleep for ~49710 years before the exception is triggered. I think this is more than enough. :-)

Reply:
Exactly. You could hit the 1ns if the operating system clock was concurrently moved backward by a NTP adjustment of ~ 1024 seconds at the same time that the offset is taken but this would be extremely unlikely - I don't believe it deserves any infinite retry loop. That was discussed at the time the enhancement that provides sub-millisecond precision was proposed, and it was rejected in favor of a one time retry.